[modelica] Initial implementation for PMD

atrosinenko opened the pull request:
This PR introduces initial support of Modelica language for CPD. It passes existing tests (although not yet adds its own). It seems to basically work when tested on the [Modelica Standard Library](https://github.com/modelica/ModelicaStandardLibrary).

Its grammar is `modelica.g4` taken from https://github.com/antlr/grammars-v4 and (textually) significantly rewritten to eliminate usage of anonymous string literals (I tried to retain all the original meaning of the grammar). Though, I'm not experienced with Antlr at all, so please feel free to point me to any silly mistakes / non-idiomatic constructs. :)

PS: Is it "correctly processed" copyright notice that reside in `Modelica.g4`? I left it intact except for adding a note that I have modified it, since I easily could introduce some mistakes not existing in the original version.

---
Modelica: https://modelica.org/modelicalanguage

> The Modelica Language is a non-proprietary, object-oriented, equation based language to conveniently model complex physical systems containing, e.g., mechanical, electrical, electronic, hydraulic, thermal, control, electric power or process-oriented subcomponents. 

Language Specification: https://specification.modelica.org/v3.4/MLS.html
Documentation: https://doc.modelica.org/

pmd-test commented:
<!--
  0 Errors
  0 Warnings
  1 Message: No java rules are changed!
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">No java rules are changed!</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

atrosinenko commented:
Hmm... Maybe it is worth to re-implement it using JavaCC, because I'm planning implementing some static analysis.

adangel commented:
Depending on how urgent you need static analysis for modelica, you might wait for PMD 7. There we have full antlr support for languages. On PMD 6 (master), we only support it with javacc (or with wrapping an external parser).

atrosinenko commented:
I have manually translated `Modelica.g4` to JavaCC and, surprisingly, it still basically works for CPD... Here is a Modelica skeleton for PMD.

oowekyala commented on pmd-modelica/etc/grammar/Modelica.jjt:
```diff
@@ -0,0 +1,952 @@
+void Equation(): {}
+{
+    (
+          (SimpleExpression() <EQ_SIGN> Expression())
```
This looks like it needs its own node, eq `RegularEquation` or whatever that is.

oowekyala commented on pmd-modelica/etc/grammar/Modelica.jjt:
```diff
@@ -0,0 +1,952 @@
+        | ForEquation()
+        | ConnectClause()
+        | WhenEquation()
+        | (Name() FunctionCallArgs())
```
Same here

oowekyala commented on pmd-modelica/etc/grammar/Modelica.jjt:
```diff
@@ -0,0 +1,952 @@
+void Statement(): {}
+{
+    (
+          (ComponentReference() ((<ASSIGN> Expression()) | FunctionCallArgs()))
```
Should be different nodes
```
ComponentReference() ((<ASSIGN> Expression()) #AssignmentStatement(2) | FunctionCallArgs() #FunctionCallStatement(2))
```

adangel commented on pmd-modelica/src/main/resources/category/modelica/misc.xml:
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0"?>
+
+<ruleset name="Misc Modelica rules"
```
Actually, there is no "misc" category. This ruleset should probably live under tests (somewhere under src/test/resources).

atrosinenko commented on pmd-modelica/etc/grammar/Modelica.jjt:
```diff
@@ -0,0 +1,952 @@
+        <ELSEIF>
+        Expression()
+        <THEN>
+        (Equation() <SC>)*
```
Looks like at least OpenModelica allows empty statement lists in if-statements, so
```modelica
if time > 10 then
end if;
```
seems to be a correct algorithm. This even seems not very weird. With equations, there are restrictions on their number (number of equations == number of state variables) but that's another story, not about grammar...

atrosinenko commented:
Yet another try... Now, it parses the attached file successfully, but has some general lookahead-related issues. The AST now seems significantly less deep. Not sure on its usability yet, though, now it is quite simple to test new `jjt` grammars, since it requires modification of two adapter implementations only.

_Notifying @JKRT_

[Test.mo.txt](https://github.com/pmd/pmd/files/3694803/Test.mo.txt)

atrosinenko commented:
Thanks very much @oowekyala for thorough explanation!

JKRT commented:
> 
> 
> Yet another try... Now, it parses the attached file successfully, but has some general lookahead-related issues. The AST now seems significantly less deep. Not sure on its usability yet, though, now it is quite simple to test new `jjt` grammars, since it requires modification of two adapter implementations only.
> 
> _Notifying @JKRT_
> 
> [Test.mo.txt](https://github.com/pmd/pmd/files/3694803/Test.mo.txt)

I must say this addition to PMD makes me really happy!

I have just eyeballed your PR, but I have been going around thinking the same thing for a time.. 

Currently, I am a bit short on time... I have subscribed to the thread and will follow up on this sometime next week

atrosinenko commented:
Now it seems to be able to parse the entire Modelica Standard Library and execute a trivial rule on it:
```xpath
//ClassSpecifier/*[SimpleName[1]/@Image != SimpleName[last()]/@Image]
```
It validates that the start and end markers of a class has the same name inside.

[This preview](https://github.com/atrosinenko/pmd/releases/tag/modelica-ast-preview) is intended for testing of AST structure by those who know Modelica better.

Inviting @sjoelund @adrpo as well, who were spotted among persons made some of the latest commits to OpenModelica parser. :)

oowekyala commented on pmd-modelica/etc/grammar/Modelica.jjt:
```diff
@@ -0,0 +1,952 @@
+    ClassModification()
+}
+
+TOKEN: {
```
It's good practice to declare all tokens together at the top of the file. Otherwise JavaCC may generate implicit tokens at the first point of use in some specific cases...

Also, related to the rest of the grammar: if the lexer is case-sensitive (which I assume yours is), you don't have to refer to tokens as `<THEN>`, you can just write `"then"`, which improves readability.

atrosinenko replied:
> Also, related to the rest of the grammar: if the lexer is case-sensitive (which I assume yours is), you don't have to refer to tokens as <THEN>, you can just write "then", which improves readability.

Not sure it is a big problem, because, on the other hand, they are easily distinguishable even without the syntax highlighting, so it looks like both approaches have their pros and cons. I see both styles from language to language in the PMD repository. I don't argue against being requested to change this, though. (The other question, maybe I misunderstood some JJTree syntax, `[ "a" ]` expressions for example, are somewhere referenced as equivalents of `("a")?`, if I get it right...)

atrosinenko replied:
> It's good practice to declare all tokens together at the top of the file. Otherwise JavaCC may generate implicit tokens at the first point of use in some specific cases...

I'm not so familiar with JavaCC yet, but if this can eliminate some obscure debugging, then I should definitely fix this. Maybe after some other review comments.

oowekyala replied:
> it looks like both approaches have their pros and cons.

Yes, ultimately that's up to your preference. I usually use syntax highlighting and prefer the less ceremonial form with quotes, but whatever. I don't expect you to change it

>` [ "a" ]` expressions for example, are somewhere referenced as equivalents of `("a")?`

Yes the form with square brackets is just sugar, sometimes it's more readable... I use those forms inconsistently myself tbh

atrosinenko replied:
I just want to say it looks like regular expression token, e.g., in `cpp.jj`:
```
<  OCTALINT : "0" (["'", "0"-"7"])* >
```
but has different meaning. So it was kind of confusing for me when I read _different_ tutorials :)

oowekyala replied:
Oh yeah right, the Javacc language has two sub-languages for regular expressions and BNF expansions, differing only in subtle ways. In a regular expression context,  `[...]` means a character class, while in BNF it's like `(...)?`. Regular expressions more complicated than just a string literal are all enclosed in `< ... >`, so you can identify the context like that. 

Side note: I think it's probably a very old mistake from the JavaCC guys to require string literals in character classes. They could have required character literals, and this would have been much less ambiguous (and for them, easier to parse).

atrosinenko replied:
> It's good practice to declare all tokens together at the top of the file.

Fixed

oowekyala commented:
> Not sure on its usability yet, though

Maybe, we can include the module in the next release under a "beta" status, so that you can make further changes to the grammar to refine it a bit before advertising it as stable

atrosinenko commented:
> Maybe, we can include the module in the next release under a "beta" status, so that you can make further changes to the grammar to refine it a bit before advertising it as stable

I supposed something similar: wait for some feedback from OpenModelica developers a bit, if any (this is a Modelica implementation I'm most familiar with), then merge it, so others could populate the rulesets using regular development process. This would allow others to "test drive" new language implementation more widely and spot issues before officially stabilizing AST.

atrosinenko commented:
Just for fun: [msl-copy-paste.txt](https://github.com/pmd/pmd/files/3701092/msl-copy-paste.txt).

At least the first instance is not a CPD implementation bug:
```
Found a 180 line (3474 tokens) duplication in the following files: 
Starting at line 1745 of /tmp/Modelica/Media/Air/ReferenceMoistAir.mo
Starting at line 1808 of /tmp/Modelica/Media/Water/IF97_Utilities.mo
```
(This is from copy-pasted `function g2 "Gibbs function for region 2: g(p,T)"`)

atrosinenko commented:
[New preview](https://github.com/atrosinenko/pmd/releases/tag/modelica-ast-preview-v2) build is published: it has Modelica syntax highlighting in the Designer and initial support for File/Class scopes and Component list.

atrosinenko commented:
This hackish flattening process definitely needs some debugging. But it passes some simple tests.

atrosinenko commented:
Squashed and rebased onto new `master`.

JKRT commented:
> 
> 
> Squashed and rebased onto new `master`. (still WIP)

I am still busy with other stuff. However, I do applaud this effort!

atrosinenko commented:
Thank you very much for suggestions!

For now, I'm more Scala developer than Java one (and so, for example, not very familiar with Java enumerations) and I'm completely new to designing **public** APIs, so the suggestions are very valuable.

Now I'm completely rewriting, simplifying and documenting the symbol resolution machinery and aligning it with the Modelica Language Specification (with some refs in javadoc). Meanwhile, is there some interface I can implement, so this information would at least show up in the Rule Designer?

oowekyala commented:
> Now I'm completely rewriting, simplifying and documenting the symbol resolution machinery and aligning it with the Modelica Language Specification (with some refs in javadoc)

Good to hear :)

> Meanwhile, is there some interface I can implement, so this information would at least show up in the Rule Designer?

The designer uses Scope and VariableNameDeclaration for now. I think, we need to abstract it away (since even the java module won't use those with 7.0). I'll give it a try

atrosinenko commented:
_This just pushed commit is merely a safepoint and still WIP._

atrosinenko commented on pmd-modelica/src/main/java/net/sourceforge/pmd/lang/modelica/symboltable/ClassRestriction.java:
```diff
@@ -0,0 +1,45 @@
+        return this == FUNCTION || this == OPERATOR_FUNCTION;
+    }
+
+    void setPure(boolean functionIsPure) {
```
Thanks! This was definitely an error, fixed.

atrosinenko commented:
Fixed most of the review issues. Small refinement of class flattening is pending, then it would be great if someone familiar with the Modelica semantics would glance at it.

I have detangled my code from the deprecated Scope and Declaration APIs but used a homegrown API for now. It abstracts various notions of Modelica with different interfaces:
* `ModelicaScope` -- a lexical scope of source file
* `ModelicaDeclaration` -- something that is explicitly declared with some name inside some scope
* `ModelicaType` -- something that can be used to specify a type of some component, either explicitly declared class or built-in
* `SubcomponentResolver` -- something that can be queried "from the outside" for further resolution (that is, not lexically)
* `ResolvableEntity` -- something that resolver _can find_ as a result of resolution process

JKRT commented:
@atrosinenko I really salute this work : ), How is the progress

atrosinenko commented:
> How is the progress

Mostly awaiting review and maybe some testing.

adangel commented on pmd-modelica/pom.xml:
```diff
@@ -0,0 +1,83 @@
+    <dependencies>
+        <dependency>
+            <groupId>org.antlr</groupId>
+            <artifactId>antlr4-runtime</artifactId>
```
This dependency can be removed, since we are using javacc.

adangel commented on pmd-modelica/src/main/java/net/sourceforge/pmd/lang/modelica/ast/ASTClassDefinition.java:
```diff
@@ -0,0 +1,78 @@
+    private ModelicaClassSpecialization specialization;
+    private ModelicaClassSpecifierNode specifier;
+
+    public ASTClassDefinition(int id) {
```
Please change these constructors to package-private.

adangel commented on pmd-modelica/src/main/java/net/sourceforge/pmd/lang/modelica/ast/ASTElementList.java:
```diff
@@ -0,0 +1,30 @@
+public class ASTElementList extends AbstractModelicaNode {
+    private Visibility visibility;
+
+    public ASTElementList(int id) {
```
Please change these constructors to package-private.

adangel commented on pmd-modelica/src/main/java/net/sourceforge/pmd/lang/modelica/ast/ASTStoredDefinition.java:
```diff
@@ -0,0 +1,55 @@
+/**
+ * A representation of a Modelica source code file.
+ */
+public class ASTStoredDefinition extends AbstractModelicaNode {
```
This seems to be the root node of the AST. Please mark this by implementing the marker interface `net.sourceforge.pmd.lang.ast.RootNode`.

adangel commented on pmd-modelica/src/main/resources/category/modelica/bestpractices.xml:
```diff
@@ -0,0 +1,53 @@
+    <rule name="ClassStartNameEqualsEndName"
+          language="modelica"
+          message="Class ends with an end clause with a different name"
+          class="net.sourceforge.pmd.lang.rule.XPathRule">
```
I think, the externalInfoUrl attribute is missing - it should be:

`externalInfoUrl="${pmd.website.baseurl}/pmd_rules_modelica_bestpractices.html#classstartnameequalsendname"`

The same applies for the other two rules here.
Note: the rule documentation will be generated and is not published yet

adangel commented on pmd-modelica/src/test/resources/net/sourceforge/pmd/lang/modelica/rule/bestpractices/xml/AmbiguousResolution.xml:
```diff
@@ -0,0 +1,86 @@
+        <source-type>modelica</source-type>
+    </test-code>
+<!--
+    <test-code>
```
You can use the attribute "regressionTest=false" to have this test just ignored instead of commenting it out. See https://pmd.github.io/latest/pmd_userdocs_extending_testing.html#test-code-attributes

adangel commented on pmd-test/src/main/resources/rule-tests_1_0_0.xsd:
```diff
@@ -50,7 +50,7 @@
                 <simpleType>
                     <restriction base="string">
-                        <pattern value="(apex|ecmascript|java|jsp|plsql|pom|vf|vm|wsdl|xml|xsl)( [0-9.]+)?"></pattern>
+                        <pattern value="(apex|ecmascript|java|jsp|modelica|plsql|pom|vf|vm|wsdl|xml|xsl)( [0-9.]+)?"></pattern>
```
Interesting change :+1: - I see, we forgot to add scala here....

atrosinenko commented:
Thanks! I have squashed my commits, rebased them onto fresh master and added two small changes:
* changed `pmd-modelica` version from 6.20 to 6.21
* in Modelica.jjt, in the line
  ```
  | <#S_ESCAPE: "\\" [ "’", "\'", "\"", "?", "\\", "a", "b", "f", "n", "r", "t", "v" ] >
  ```
  changed literal Unicode character `’` to `\u2019` (expecting this change will fix Windows build)

On mergeability: since it will be experimental anyway, I think it can be merged now and then improved by subsequent commits (provided Windows build will be OK).
